### PR TITLE
Form.onChange should be called after field has changed

### LINKED ISSUE #86336: DropdownButtonFormField changes not absorbing Form onChanged Function
I tried to get Form field if any changes happen then run few operations. But, DropdownButtonFormField changes not listen Form onchanged funciton.

```
Form(
onChanged: () {
print('change Happned');
},
key: _formKey,
child: DropdownButtonFormField<String>(
                        isExpanded: true,
                        decoration: InputDecoration(
                            labelText: 'Label',
                            isDense: true,
                            contentPadding:
                                EdgeInsets.symmetric(
                                    vertical: 6.0)),
                        value: _emailAddressLabelValue,
                        focusNode: _emailAddressLabelNode,
                        hint: Text('Label'),
                        items: <String>[
                          'Home',
                          'Work',
                          'Other'
                        ]
                            .map((label) => DropdownMenuItem(
                                  child: Text(
                                    label,
                                    overflow:
                                        TextOverflow.ellipsis,
                                  ),
                                  value: label,
                                ))
                            .toList(),
                        onChanged: (value) {
                                setState(() {
                                  _emailAddressLabelValue =
                                      value!;
                                });
                              },
                      )
)
```

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #86336

[As per the docs ](https://api.flutter.dev/flutter/widgets/Form/onChanged.html)

> Called when one of the form fields changes.
In addition to this callback being invoked, all the form fields themselves will rebuild.

And [here](https://github.com/flutter/flutter/blob/738ce43d97218e0d2d4c4d5f5ecbeec39f0dbef1/packages/flutter/lib/src/widgets/form.dart#L169-L171)
```dart
  // Called when a form field has changed. This will cause all form fields
  // to rebuild, useful if form fields have interdependencies.
  void _fieldDidChange() {
```

Consider the documentation and actual usage, it is more reasonable to call the `Form.onChange` after the field changes.

### CODE REVIEW

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -1715,9 +1715,9 @@ class _DropdownButtonFormFieldState<T> extends FormFieldState<T> {
 
   @override
   void didChange(T? value) {
-    super.didChange(value);
     assert(widget.onChanged != null);
     widget.onChanged!(value);
+    super.didChange(value);
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/dropdown.dart:1720]: It is the same  for _CupertinoTextFormFieldRowState and _TextFormFieldState

Comment (xu-baolin) [reply packages/flutter/lib/src/material/dropdown.dart:1720]: Nice caught.

File: packages/flutter/test/cupertino/text_form_field_row_test.dart [hunk 2]
```diff
@@ -339,7 +369,7 @@ void main() {
           child: CupertinoTextFormFieldRow(
             onChanged: (String value) {
               _called = true;
-              expect(value, state.value);
+              changeValue = value;
```
Comment (xu-baolin) [inline packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: I am a little worried that this will be a breaking change.

Comment (xu-baolin) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]:
```dart
  void didChange(T? value) {
    setState(() {
      _value = value;
      _hasInteractedByUser.value = true;
    });
    Form.of(context)?._fieldDidChange();
  }
```
the state value and the `Form.onChange` update together.

Comment (chunhtai) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]:
```dart
void onChangedHandler(String value) {
              field.didChange(value);
              if (onChanged != null) {
                onChanged(value);
              }
            }
```
shouldn't the didChange called first before this onchange is called?

Comment (xu-baolin) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: This patch aims to call this onchange before didchange.

Comment (chunhtai) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: ah so you change this line, too?

I thought this pr is aiming at the Form.onchange is called after the field value changes, not the CupertinoTextFormFieldRow.onChange? or am I missing something?

Comment (xu-baolin) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: I really want to adjust it after onchange . Do you think this is valuable?

Comment (chunhtai) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: If you move onChanged to the front,  the formFieldState.value will not be updated before the onChanged is called. I am more toward to keep it as is

Comment (xu-baolin) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:372]: You're right, I will double check is there a way that set value first and then call onchange, and call form onchange at last.

**Comment by xu-baolin** [comment]
Close as did not found a better solution.

**Pull request closed without merging**